File: services/shop-backend/src/main/java/io/miragon/shop/adapter/outbound/persistence/cart/CartPersistenceAdapter.java
package io.miragon.shop.adapter.outbound.persistence.cart;

import io.miragon.shop.application.port.outbound.CartRepository;
import io.miragon.shop.domain.cart.Cart;
import io.miragon.shop.domain.shared.UserId;
import org.springframework.stereotype.Component;

@Component
public class CartPersistenceAdapter implements CartRepository {

    private final CartJpaRepository cartJpaRepository;

    public CartPersistenceAdapter(CartJpaRepository cartJpaRepository) {
        this.cartJpaRepository = cartJpaRepository;
    }

    @Override
    public Cart loadCart(UserId userId) {
        var entity = cartJpaRepository.findByUserId(userId.getValue());
        return entity != null ? CartMapper.toDomain(entity) : new Cart(userId, java.util.List.of());
    }

    @Override
    public void storeCart(Cart cart) {
        var entity = CartMapper.toEntity(cart);
        cartJpaRepository.save(entity);
    }

    @Override
    public void deleteCart(UserId userId) {
        cartJpaRepository.deleteById(userId.getValue());
    }
}

File: services/shop/shop-backend/src/main/kotlin/io/miragon/shop/adapter/outbound/persistence/cart/CartPersistenceAdapter.kt
package io.miragon.shop.adapter.outbound.persistence.cart

import io.miragon.shop.adapter.outbound.persistence.cart.CartMapper.toDomain
import io.miragon.shop.adapter.outbound.persistence.cart.CartMapper.toEntity
import io.miragon.shop.application.port.outbound.CartRepository
import io.miragon.shop.domain.cart.Cart
import io.miragon.shop.domain.shared.UserId
import org.springframework.stereotype.Component

@Component
class CartPersistenceAdapter(
    private val cartJpaRepository: CartJpaRepository
) : CartRepository {

    override fun loadCart(userId: UserId): Cart {
        val cartEntity = cartJpaRepository.findByUserId(userId.value)
        return cartEntity?.toDomain() ?: Cart(userId = userId, items = emptyList())
    }

    override fun storeCart(cart: Cart) {
        val cartEntity = CartEntity.toEntity(cart)
        cartJpaRepository.save(cartEntity)
    }

    override fun deleteCart(userId: UserId) {
        cartJpaRepository.deleteById(userId.value)
    }
}